File: 01094. Car Pooling/solution.go
/**
 * Problem: 1094. Car Pooling
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func carPooling(trips [][]int, capacity int) bool {
	// Initialize an array to track the net change in passengers at each location.
	// The size is 1001 because locations range from 0 to 1000.
	passengerChange := make([]int, 1001)

	// Iterate over each trip to record the pickup and drop-off changes.
	for i := 0; i < len(trips); i++ {
		// Add passengers at the start location.
		passengerChange[trips[i][1]] += trips[i][0]
		// Remove passengers at the end location.
		passengerChange[trips[i][2]] -= trips[i][0]
	}

	// Initialize the available seats with the given capacity.
	availableSeats := capacity

	// Traverse all possible locations to simulate the car's journey.
	for i := 0; i < len(passengerChange); i++ {
		// Update available seats based on passenger changes at the current location.
		availableSeats -= passengerChange[i]

		// If available seats go below zero, capacity is exceeded.
		if availableSeats < 0 {
			return false
		}
	}

	// If the journey completes without exceeding capacity, return true.
	return true
}

File: 01094. Car Pooling/solution.js
/**
 * Problem: 1094. Car Pooling
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Checks if trips fit within car capacity
 *
 * @param {number[][]} trips - Each trip as [passengers, from, to]
 * @param {number} capacity - Car's max capacity
 *
 * @returns {boolean} - True if possible, else false
 */
const carPooling = (trips, capacity) => {
  // Sort trips by start location for processing pickups
  const sortedByStart = [...trips].sort((a, b) => a[1] - b[1])
  // Sort trips by end location for processing drop-offs
  const sortedByEnd = [...trips].sort((a, b) => a[2] - b[2])

  // Track the current number of passengers in the car
  let currentPassengerCount = 0

  // Use two pointers to process pickups and drop-offs
  for (
    let pickupIdx = 0, dropoffIdx = 0;
    pickupIdx < trips.length;
    pickupIdx++
  ) {
    // Drop off passengers whose trip has ended before the current pickup
    while (sortedByStart[pickupIdx][1] >= sortedByEnd[dropoffIdx][2]) {
      currentPassengerCount -= sortedByEnd[dropoffIdx][0] // Remove passengers
      dropoffIdx++ // Move to next drop-off
    }

    // Pick up passengers for the current trip
    currentPassengerCount += sortedByStart[pickupIdx][0]

    // If capacity exceeded, return false
    if (currentPassengerCount > capacity) return false
  }

  // All trips processed without exceeding capacity
  return true
}
